src/ring_theory/valuation/valuation_ring.lean

```diff
@@ -270,6 +278,135 @@ end
+  exact associated_of_dvd_dvd (irreducible.dvd_symm hq hp this) this,
+end
+
+lemma iff_is_integer_or_is_integer :
+  valuation_ring R ↔ ∀ x : K, is_localization.is_integer R x ∨ is_localization.is_integer R x⁻¹ :=
+begin
+  split,
+  { introsI H x,
+    obtain ⟨x : R, y, hy, rfl⟩ := is_fraction_ring.div_surjective x,
+    any_goals { apply_instance },
+    have := (map_ne_zero_iff _ (is_fraction_ring.injective R K)).mpr (non_zero_divisors.ne_zero hy),
+    obtain ⟨s, rfl|rfl⟩ := valuation_ring.cond x y,
+    { exact or.inr ⟨s, eq_inv_of_mul_eq_one_left $
+        by rwa [mul_div, div_eq_one_iff_eq, map_mul, mul_comm]⟩ },
+    { exact or.inl ⟨s, by rwa [eq_div_iff, map_mul, mul_comm]⟩ } },
+  { intro H,
+    constructor,
+    intros a b,
+    by_cases ha : a = 0, { subst ha, exact ⟨0, or.inr $ mul_zero b⟩ },
+    by_cases hb : b = 0, { subst hb, exact ⟨0, or.inl $ mul_zero a⟩ },
+    replace ha := (map_ne_zero_iff _ (is_fraction_ring.injective R K)).mpr ha,
+    replace hb := (map_ne_zero_iff _ (is_fraction_ring.injective R K)).mpr hb,
+    obtain ⟨c, e⟩|⟨c, e⟩ := H (algebra_map R K a / algebra_map R K b),
+    { rw [eq_div_iff hb, ← map_mul, (is_fraction_ring.injective R K).eq_iff, mul_comm] at e,
+      exact ⟨c, or.inr e⟩ },
+    { rw [inv_div, eq_div_iff ha, ← map_mul,
+        (is_fraction_ring.injective R K).eq_iff, mul_comm c] at e,
+      exact ⟨c, or.inl e⟩ } }
+end
+
+lemma is_integer_or_is_integer [h : valuation_ring R] (x : K) :
+  is_localization.is_integer R x ∨ is_localization.is_integer R x⁻¹ :=
+iff_is_integer_or_is_integer.mp h x
+
+@[priority 100]
+instance [valuation_ring R] : is_bezout R :=
+begin
+  classical,
+  rw is_bezout.iff_span_pair_is_principal,
+  intros x y,
+  rw ideal.span_insert,
+  cases le_total (ideal.span {x} : ideal R) (ideal.span {y}),
+  { erw sup_eq_right.mpr h, exact ⟨⟨_, rfl⟩⟩ },
+  { erw sup_eq_left.mpr h, exact ⟨⟨_, rfl⟩⟩ }
+end
+
+lemma iff_local_bezout_domain :
+  valuation_ring R ↔ local_ring R ∧ is_bezout R :=
+begin
+  classical,
+  refine ⟨λ H, by exactI ⟨infer_instance, infer_instance⟩, _⟩,
+  rintro ⟨h₁, h₂⟩,
+  resetI,
+  refine iff_dvd_total.mpr ⟨λ a b, _⟩,
+  obtain ⟨g, e : _ = ideal.span _⟩ := is_bezout.span_pair_is_principal a b,
+  obtain ⟨a, rfl⟩ := ideal.mem_span_singleton'.mp
+    (show a ∈ ideal.span {g}, by { rw [← e], exact ideal.subset_span (by simp) }),
+  obtain ⟨b, rfl⟩ := ideal.mem_span_singleton'.mp
+    (show b ∈ ideal.span {g}, by { rw [← e], exact ideal.subset_span (by simp) }),
+  obtain ⟨x, y, e'⟩ := ideal.mem_span_pair.mp
+    (show g ∈ ideal.span {a * g, b * g}, by { rw e, exact ideal.subset_span (by simp) }),
+  cases eq_or_ne g 0 with h h, { simp [h] },
+  have : x * a + y * b = 1,
+  { apply mul_left_injective₀ h, convert e'; ring_nf },
+  cases local_ring.is_unit_or_is_unit_of_add_one this with h' h',
+  left, swap, right,
+  all_goals
+  { exact mul_dvd_mul_right (is_unit_iff_forall_dvd.mp (is_unit_of_mul_is_unit_right h') _) _ },
+end
+
+protected lemma tfae :
```

Comment:
Again, I think it would be better if `R` and `K` are both explicit for this lemma.
Also, can you add a tfae lemma without `∀ x : K, is_localization.is_integer R x ∨ is_localization.is_integer R x⁻¹`, so we have something without the need for a fraction field around?

Reply:
I changed the statement to be `∀ x : fraction_ring R`. I don't think we need a separate `tfae` that allows a different fraction field?

Reply:
Sounds good to me!